### PR TITLE
fix: required check failed on ref, ref_array, ontology, ontology_array

### LINKED ISSUE #1294: Mandatory column constraint not working in RepeatedVariables
**What version of EMX2 are you using (see footer)**
8.71.0

**Describe the bug**
I can upload RepeatedSourceVariables with empty column isRepeatOf.name and/or isRepeatOf.table, while RepeatedSourceVariables.isRepeatOf is mandatory e.g. see RepeatedSourceVariables sheet in this data: 
[test1652954588281.xlsx](https://github.com/molgenis/molgenis-emx2/files/8728113/test1652954588281.xlsx)

image

**To Reproduce**
1. Go to 'https://data-catalogue-staging-test.molgeniscloud.org/'
2. Make a new schema
3. Upload example test file

### PR DESCRIPTION
**Pull request by mswertz**
this was consequence when we added 'validation' that required checks were removed from postgresql and moved to java.
setRequired() was not correctly passed to the technical columns for insert/update.

fixes #1294

### CODE REVIEW

File: backend/molgenis-emx2/src/main/java/org/molgenis/emx2/Row.java [hunk 1]
```diff
@@ -358,6 +358,9 @@ public boolean notNull(String columnName) {
   }
 
   public boolean isNull(String columnName, ColumnType type) {
+    if (type.isArray()) {
```
Comment (mswertz) [inline backend/molgenis-emx2/src/main/java/org/molgenis/emx2/Row.java:361]: this is part of the actual fix

File: backend/molgenis-emx2/src/main/java/org/molgenis/emx2/TableMetadata.java [hunk 2]
```diff
@@ -220,7 +220,9 @@ public List<Column> getExpandedColumns(List<Column> columns) {
           if (!ref.isOverlapping()) { // only add overlapping once
             result.put(
                 ref.getName(),
-                new Column(c.getTable(), ref.getName(), true).setType(ref.getPrimitiveType()));
+                new Column(c.getTable(), ref.getName(), true)
+                    .setType(ref.getPrimitiveType())
+                    .setRequired(c.isRequired()));
```
Comment (mswertz) [inline backend/molgenis-emx2/src/main/java/org/molgenis/emx2/TableMetadata.java:225]: and in particular this setRequired was causing the bug

File: backend/molgenis-emx2/src/main/java/org/molgenis/emx2/Row.java [hunk 3]
```diff
@@ -358,6 +358,9 @@ public boolean notNull(String columnName) {
...
+      return get(columnName, type) == null || ((Object[]) get(columnName, type)).length == 0;
```
Comment (connoratrug) [inline backend/molgenis-emx2/src/main/java/org/molgenis/emx2/Row.java:362]: "get"  What ? getColumnValue ?? , why would i need to supply the type 
the method is called isNull , but checks if null or empty array, and i looks like 'get' returns null in case of array but its hard to see what is going in due to the use of single char variable names

File: backend/molgenis-emx2/src/main/java/org/molgenis/emx2/TableMetadata.java (hunk 2 above)
Comment (connoratrug) [reply backend/molgenis-emx2/src/main/java/org/molgenis/emx2/TableMetadata.java:225]: suggest to move creation of the expanded ref column to a named function , hard to spot if additional properties need to be copied , could also be a function on the Reference object

File: backend/molgenis-emx2/src/main/java/org/molgenis/emx2/Row.java (hunk 3 above)
Comment (mswertz) [reply backend/molgenis-emx2/src/main/java/org/molgenis/emx2/Row.java:362]: ? single var variable names?

Comment (mswertz) [reply backend/molgenis-emx2/src/main/java/org/molgenis/emx2/Row.java:362]: N.B. the row cannot know what type you want to get() the data. So you either can use typed getters, e.g. getString, getInt etc, or you must parameterize your request. For example, if you are currently parsing a CSV file then the underlying value might a comma seperated string that first needs parsing before you can assess nullable. Now, if we think hard about all edge cases we might be able to asses 'nullOrEmpty'-ness based on the underlying this.values.get(columnName) which might be null, "", [] without using columnType.

Basically I propose to rename to 'isNullOrEmpty' to clarify there is a bit more to this.

**Review by connoratrug — APPROVED** [review]
